src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -14,43 +14,142 @@ internal static class CompletionHelpers
         {
+            matchStrategy ??= DefaultMatch;
+
             string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
-            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
+            if (quote != SingleQuote)
+            {
+                wordToComplete = NormalizeToExpandableString(wordToComplete);
+            }
 
             foreach (string value in possibleCompletionValues)
             {
-                if (pattern.IsMatch(value))
+                if (matchStrategy(value, wordToComplete))
                 {
                     string completionText = QuoteCompletionText(value, quote);
+                    string toolTip = toolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = listItemTextMapping?.Invoke(value) ?? value;
 
-                    string listItemText = value;
-
-                    yield return new CompletionResult(
-                        completionText,
-                        listItemText,
-                        resultType,
-                        toolTip: toolTipMapping is null
-                            ? listItemText
-                            : toolTipMapping(value));
+                    yield return new CompletionResult(completionText, listItemText, resultType, toolTip);
                 }
             }
         }
 
+        /// <summary>
+        /// Normalizes the input string to an expandable string format for PowerShell.
+        /// </summary>
+        /// <param name="value">The input string to be normalized.</param>
+        /// <returns>The normalized string with special characters replaced by their PowerShell escape sequences.</returns>
+        /// <remarks>
+        /// This method replaces special characters in the input string with their PowerShell equivalent escape sequences:
+        /// <list type="bullet">
+        ///     <item><description>Replaces "\r" (carriage return) with "`r".</description></item>
+        ///     <item><description>Replaces "\n" (newline) with "`n".</description></item>
+        ///     <item><description>Replaces "\t" (tab) with "`t".</description></item>
+        ///     <item><description>Replaces "\0" (null) with "`0".</description></item>
+        ///     <item><description>Replaces "\a" (bell) with "`a".</description></item>
+        ///     <item><description>Replaces "\b" (backspace) with "`b".</description></item>
+        ///     <item><description>Replaces "\u001b" (escape character) with "`e".</description></item>
+        ///     <item><description>Replaces "\f" (form feed) with "`f".</description></item>
+        ///     <item><description>Replaces "\v" (vertical tab) with "`v".</description></item>
+        /// </list>
+        /// </remarks>
+        internal static string NormalizeToExpandableString(string value)
+            => value
+                .Replace("\r", "`r")
+                .Replace("\n", "`n")
+                .Replace("\t", "`t")
```

Comment:
The method could be optimized for performance as separate work (low priority).

Reply:
This one is actually quite difficult to optimize, basic `Replace` approach outperforms other methods (string builder, regex etc.).